homeassistant/components/timer.py

```diff
@@ -0,0 +1,402 @@
+                        minutes=service.data.get(
+                            ATTR_MINUTES, DEFAULT_DURATION),
+                        seconds=service.data.get(
+                            ATTR_SECONDS, DEFAULT_DURATION),
+                    )
+                )
+        if tasks:
+            yield from asyncio.wait(tasks, loop=hass.loop)
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml')
+    )
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_START, async_handler_service,
+        descriptions[DOMAIN][SERVICE_START], SERVICE_SCHEMA_DURATION)
+    hass.services.async_register(
+        DOMAIN, SERVICE_PAUSE, async_handler_service,
+        descriptions[DOMAIN][SERVICE_PAUSE], SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_CANCEL, async_handler_service,
+        descriptions[DOMAIN][SERVICE_CANCEL], SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_FINISH, async_handler_service,
+        descriptions[DOMAIN][SERVICE_FINISH], SERVICE_SCHEMA)
+
+    yield from component.async_add_entities(entities)
+    return True
+
+
+class Timer(Entity):
+    """Representation of a timer."""
+
+    def __init__(self, hass, object_id, name, icon,
+                 weeks, days, hours, minutes, seconds):
+        """Initialize a timer."""
+        self.entity_id = ENTITY_ID_FORMAT.format(object_id)
+        self._name = name
+        self._state = STATUS_IDLE
+        self._duration = timedelta(weeks=weeks, days=days, hours=hours,
+                                   minutes=minutes, seconds=seconds)
+        self._remaining = self._duration
+        self._icon = icon
+        self._hass = hass
+        self._start = None
+        self._end = None
+        self._listener = None
+
+    @property
+    def should_poll(self):
+        """If entity should be polled."""
+        return True
+
+    @property
+    def name(self):
+        """Return name of the timer."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to be used for this entity."""
+        return self._icon
+
+    @property
+    def state(self):
+        """Return the current value of the timer."""
+        return self._state
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        start = dt_util.as_local(self._start) if self._start else None
+        end = dt_util.as_local(self._end) if self._end else None
+        return {
+            ATTR_DURATION: self._duration.__str__(),
+            ATTR_START: start,
+            ATTR_END: end,
+            ATTR_REMAINING: self._remaining.__str__()
+        }
+
+    @Throttle(UPDATE_INTERVAL)
+    def update(self):
+        """Calculate remaining time."""
+        if self._state == STATUS_ACTIVE:
+            self._remaining = self._end - dt_util.utcnow()
```

Comment:
We should never ever store a state or state attribute that is time based. That's a) causing unnecessary updates and b) means that the moment it is written it's out dated. For the frontend we should just calculate the current time that is remaining.

Reply:
As asked above: this can be done in UI? I added the update so people get an estimate of the remaining time when looking at it.

Reply:
It's not needed and it's also against the design of HASS. A state or related attributes should not change based on time. (time left, % of video that is currently played etc)